text/0000-multi-type-return-position-impl-trait.md

```diff
@@ -0,0 +1,673 @@
+
+## (TODO) Representing `Self` in traits and types
+
+TODO (tldr: when generating types which expect a `Self`, make sure to wrap the
+output in the appropriate enum member branch. If anyone later asks what type it
+is: just say it's an `impl Trait`.)
+
+## (TODO) `Any` trait and `TypeId`
+
+TODO (tldr: proxy the type id to the inner type, that's what `dyn` does too. You
+can't downcast `impl Trait` as-is anyway, so it's likely fine).
+
+# Interaction with lifetimes
+
+`dyn Trait` already supports multi-type _dynamic_ dispatch. The rules we're
+proposing for multi-type _static_ dispatch using `impl Trait` should mirror the
+existing rules we apply to `dyn Trait.` We should follow the same lifetime rules
+for multi-type `impl Trait` as we do for `dyn Trait`:
+
+```rust
+fn multi_iter<'a>(x: i32, iter_a: &'a mut std::ops::Range<i32>) -> impl Iterator<Item = i32> + 'a {
+    match x {
+        0 => iter_a,                  // `&'a std::ops::Range<i32>`
+        _ => vec![5, 10].into_iter(), // `std::vec::IntoIter<i32>`
+    }
+}
+```
+
+This code should be desugared by the compiler into something resembling the following
+([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=60ddacbb20c4068a0fff44a5481a7136)):
+
+```rust
+enum Enum<'a> {
+    A(&'a mut std::ops::Range<i32>),
+    B(std::vec::IntoIter<i32>),
+}
+
+impl<'a> Iterator for Enum<'a> {
+    type Item = i32;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Enum::A(iter) => iter.next(),
+            Enum::B(iter) => iter.next(),
+        }
+    }
+
+    // ..repeat for the remaining 74 `Iterator` trait methods
+}
+
+fn multi_iter<'a>(x: i32, iter_a: &'a mut std::ops::Range<i32>) -> Enum<'a> {
+    match x {
+        0 => Enum::A(iter_a),
+        _ => Enum::B(vec![5, 10].into_iter()),
+    }
+}
+```
+
+It should be fine if multiple iterators use the same lifetime. But only a single
+lifetime should be permitted on the return type, as is the case today when
+using `dyn Trait`:
+
+```rust
+// ❌ Fails to compile (Rust 1.66.0)
+// error[E0226]: only a single explicit lifetime bound is permitted
+fn fails<'a, 'b>() -> Box<dyn Iterator + 'a + 'b> {
+    ...
+}
+```
+
+# (TODO) Relationship to `dyn`
+
+TODO (explain the relationship to dyn traits, returning `Self`, and which traits
+can and can't be allowed, and how that differs from single-value TAITs).
+
+## (TODO) Parity in lifetime rules
+
+TODO: Move lifetimes section here
+
+## (TODO) Sized and Self
+
+TODO (tldr: neither is an issue with `impl Trait` today, and it shouldn't have
+to be either in the future. That's the main difference in this approach.)
+
+# Drawbacks
+
+## (TODO) Downcasting
+
+TODO (tldr: because we're generating an anonymous enum, you can't downcast to a
+concrete type. With single-value RPIT you have a concrete type which you _can_
+downcast to. But with MTRPIT we're generating a type, meaning you can't downcast
+into it)
```

Comment:
This feels like something the compiler could do? It knows about what types are in the enum so theoretically it should be able to check if any type in the enum is the downcast target and delegate the downcast?

Should this be under "Future Possibilities"?

Reply:
`Any::downcast` and friends downcast to whatever `Any::type_id` says. `Any` is blanket implemented for all `Sized + 'static` types; thus without specialization a dispatch enum will downcast to the dispatch enum, and downcast specialization would be extremely unsound (as `Any` downcast is a reinterpret cast including for owned memory).

Returning a trait object `dyn Any + 'static` will downcast to the individual hidden types, because the type id used is the one from the vtable of the hidden type.

Other traits in the ecosystem exist that use `downcast(&self)` methods or the [provider api](https://doc.rust-lang.org/stable/std/any/trait.Provider.html#); those will of course dispatch properly.

Reply:
I'm not convinced that is an impossible issue.

Keep in mind that the desugaring here is shown for simplicity, not for optimization.

Whatever vtable-like thing is used for the dispatch (be that an enum or on-stack vtable) could hold the typeids.

Reply:
By reference, it could work, so long as the hidden type is at offset 0.

By stack value, it could maybe work, as you would need to use `transmute_prefix` instead of `transmute` because of the differing size.

The problem is the differing size once it's heap allocated behind a box. `Box<dyn Any + 'static>::downcast<T>` gives you `Box<T>` via a reinterpret cast. This is guaranteed; the way to write `downcast_unchecked` is `Box::from_raw(Box::into_raw(it) as *mut T)`. A box with a different layout than the allocation is UB, realized when deallocating with a different layout than used for allocating.

It *can* work if you have `MagicInline<dyn Any>` which *dereferences* to the `dyn Any` of the hidden type, but that's because of the [same container caveat](https://doc.rust-lang.org/stable/std/any/index.html#smart-pointers-and-dyn-any) that exists for every other container, that `.type_id()` will give you the type id of the container, not of the contained type. The container is also `impl Any` separately from the contained.